go/ipfs/options.go

```diff
@@ -0,0 +1,107 @@
+package ipfs
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/ipfs/go-ipfs/repo"
+	"github.com/pkg/errors"
+)
+
+// resetRepoConfigPorts adds portIdx to each default port number. The index is added instead of replaced to keep
+// idx 0 and 1 from conflicting.
+func resetRepoConfigPorts(r repo.Repo, portIdx int) error {
+	apiPort := strconv.Itoa(DefaultAPIPort + portIdx)
+	gatewayPort := strconv.Itoa(DefaultGatewayPort + portIdx)
+
+	swarmPort := strconv.Itoa(DefaultSwarmPort + portIdx)
+
+	rc, err := r.Config()
+	if err != nil {
+		return errors.Wrap(err, "error getting IPFS repo config")
+	}
+
+	rc.Addresses.API = strings.Replace(rc.Addresses.API, "5001", apiPort, -1)
+	rc.Addresses.Gateway = strings.Replace(rc.Addresses.Gateway, "8080", gatewayPort, -1)
+	for i, addr := range rc.Addresses.Swarm {
+		rc.Addresses.Swarm[i] = strings.Replace(addr, "4001", swarmPort, -1)
+	}
+
+	return errors.Wrap(r.SetConfig(rc), "error setting IPFS repo config")
+}
+
+// An Option configures the IPFS ChunkStore
+type Option interface {
+	apply(*config) error
+}
+
+type optFunc func(*config) error
+
+func (of optFunc) apply(c *config) error {
+	return of(c)
+}
+
+// SetLocal makes the ChunkStore only use the local IPFS blockstore for both reads and writes.
+func SetLocal() Option {
```

Comment:
Why the distinction between `optFunc` and `Option`? Why not just return `optFunc`?

Reply:
Also, why not:

```
func SetLocal(c *config) error {
...
}
```

...rather than returning a new function?

Reply:
`Option` should actually be refactored into `type Option func(*config) error` as I didn't end up needing anything but plain 'ol functions for the options.

All the option-setting functions return `Option` to have them all appear under the `Option` type in the documentation; see e.g. [zap's docs](https://godoc.org/go.uber.org/zap#Option) for an example